homeassistant/components/freebox/router.py

```diff
@@ -0,0 +1,199 @@
+SCAN_INTERVAL = timedelta(seconds=30)
+
+
+class FreeboxRouter:
+    """Representation of a Freebox router."""
+
+    def __init__(self, hass: HomeAssistantType, entry: ConfigEntry) -> None:
+        """Initialize a Freebox router."""
+        self.hass = hass
+        self._entry = entry
+        self._host = entry.data[CONF_HOST]
+        self._port = entry.data[CONF_PORT]
+
+        self._api: Freepybox = None
+        self._name = None
+        self.mac = None
+        self._sw_v = None
+        self._attrs = {}
+
+        self.devices: Dict[str, any] = {}
+        self.sensors: Dict[str, any] = {}
+
+        self.listeners = []
+
+    async def setup(self) -> None:
+        """Set up a Freebox router."""
+        self._api = await get_api(self.hass, self._host)
+
+        try:
+            await self._api.open(self._host, self._port)
+        except HttpRequestError:
+            _LOGGER.exception("Failed to connect to Freebox")
+            return ConfigEntryNotReady
+
+        # System
+        fbx_config = await self._api.system.get_config()
+        self.mac = fbx_config["mac"]
+        self._name = fbx_config["model_info"]["pretty_name"]
+        self._sw_v = fbx_config["firmware_version"]
+
+        # Devices & sensors
+        await self.update_all()
+        async_track_time_interval(self.hass, self.update_all, SCAN_INTERVAL)
+
+    async def update_all(self, now: Optional[datetime] = None) -> None:
+        """Update all Freebox platforms."""
+        await self.update_sensors()
+        await self.update_devices()
+
+    async def update_devices(self) -> None:
+        """Update Freebox devices."""
+        new_device = False
+        fbx_devices: Dict[str, any] = await self._api.lan.get_hosts_list()
+
+        # Adds the Freebox itself
+        fbx_devices.append(
+            {
+                "primary_name": self._name,
+                "l2ident": {"id": self.mac},
+                "vendor_name": "Freebox SAS",
+                "host_type": "router",
+                "active": True,
+                "attrs": self._attrs,
+            }
+        )
+
+        for fbx_device in fbx_devices:
+            device_mac = fbx_device["l2ident"]["id"]
+
+            if self.devices.get(device_mac) is None:
+                new_device = True
+
+            self.devices[device_mac] = fbx_device
+
+        async_dispatcher_send(self.hass, self.signal_device_update)
+
+        if new_device:
+            async_dispatcher_send(self.hass, self.signal_device_new)
+
+    async def update_sensors(self) -> None:
+        """Update Freebox sensors."""
+        # System sensors
+        syst_datas: Dict[str, any] = await self._api.system.get_config()
+
+        # According to the doc `syst_datas["sensors"]` is temperature sensors in celsius degree.
+        # Name and id of sensors may vary under Freebox devices.
+        for sensor in syst_datas["sensors"]:
+            self.sensors[sensor["id"]] = {
+                **TEMPERATURE_SENSOR_TEMPLATE,
```

Comment:
`TEMPERATURE_SENSOR_TEMPLATE` belongs to the sensor entity implementation so the router module shouldn't be aware of this.

Reply:
You just told me to do the other way around : https://github.com/home-assistant/core/pull/30334#discussion_r389965106

I am just saving a dict here, not an entity.

I don't know what to do if you don't want sensor data in the router and do not make request in the sensor platform ...

3 times I am modifying this same thing.

I want help here.

Reply:
My comment was about not doing the same work twice, here in the router module and in the sensor platform, ie not calling the device system twice.

But it's still important to separate concerns.

Here in the module we should save the device info we need later in the platforms. In the platforms we use the router device info and complement it as needed using the platform entity info, eg `TEMPERATURE_SENSOR_TEMPLATE`.

We should separate the device interface and the entity interface. The entity interface should be contained within its platform. To make that more clear, I suggest we move the entity constants, `TEMPERATURE_SENSOR_TEMPLATE` etc to the respective platforms. They are not meant to be used outside.

Reply:
Something came up to my head : is this commit ae6223ceb79bbbfe85f75c16bc9875d9a42b87af right ?
And the PR ?
And the world ? 🤪

Hope it can make it to 0.107.0